metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/storage/BlockStorage.scala

```diff
@@ -0,0 +1,190 @@
+package io.iohk.metronome.hotstuff.service.storage
+
+import cats.implicits._
+import io.iohk.metronome.storage.{KVStore, KVStoreRead, KVCollection}
+import io.iohk.metronome.hotstuff.consensus.basic.{Agreement, Block}
+import scala.collection.immutable.Queue
+
+/** Storage for blocks that maintains parent-child relationships as well,
+  * to facilitate tree traversal and pruning.
+  *
+  * It is assumed that the application maintains some pointers into the tree
+  * where it can start traversing from, e.g. the last Commit Quorum Certificate
+  * would point at a block hash which would serve as the entry point.
+  */
+class BlockStorage[N, A <: Agreement: Block](
+    blockColl: KVCollection[N, A#Hash, A#Block],
+    childToParentColl: KVCollection[N, A#Hash, A#Hash],
+    parentToChildrenColl: KVCollection[N, A#Hash, Set[A#Hash]]
+) {
+  private implicit val kvn  = KVStore.instance[N]
+  private implicit val kvrn = KVStoreRead.instance[N]
+
+  /** Insert a block into the store, and if the parent still exists,
+    * then add this block to its children.
```

Comment:
`canDelete` performs checks that disallow turning the tree into the forest. But this method allows us to put an orphaned block. Should we perform a check for parent here? Or is a forest allowed, and if so, in what situations would that be useful?

Reply:
Good observation! That is correct, it doesn't prevent the caller from inserting a forest. 

As the main docstring says, the store assumes that the the caller stores at least one pointer to the tree. If you insert independent nodes and store pointers to it, that's fine as you can still traverse everything you inserted. Maybe you wanted to store two trees, like the Prism voter chains, they each have a different genesis block.

What the store is trying to avoid is that we accidentally delete something and make parts of the tree untraversable, so the caller should trim the trees from the edges. 

I mainly wanted to avoid leaking space during pruning due to some mistake. 

In theory the store could store a pointer to the root, and disallow insertions that don't have a parent after a root has been established, but I thought since the quorum certificates store block hashes anyway, it would just slow things down and be more complicated.